lock/lock.go

```diff
@@ -0,0 +1,334 @@
+
+	return acquireLockWithRetries(options, client)
+}
+
+// acquireLockWithRetries will attempt to acquire the lock defined by the provided lock string in the configured lock table
+// for the configured region. This will retry on failure, until reaching a timeout
+func acquireLockWithRetries(options *Options, client *dynamodb.DynamoDB) error {
+	return retry.DoWithRetry(
+		options.Logger,
+		fmt.Sprintf("Trying to acquire DynamoDB lock %s in table %s\n", options.LockString, options.LockTable),
+		options.MaxRetries,
+		options.SleepBetweenRetries,
+		func() error {
+			return acquireLock(options, client)
+		},
+	)
+}
+
+// acquireLock will attempt to acquire the lock defined by the provided lock string in the configured lock table for the
+// configured region
+func acquireLock(options *Options, client *dynamodb.DynamoDB) error {
+	options.Logger.Infof("Attempting to acquire lock %s in table %s in region %s\n",
+		options.LockString,
+		options.LockTable,
+		options.AwsRegion,
+	)
+
+	putParams := &dynamodb.PutItemInput{
+		Item: map[string]*dynamodb.AttributeValue{
+			attributeLockId: {S: aws.String(options.LockString)},
+		},
+		TableName:           aws.String(options.LockTable),
+		ConditionExpression: aws.String(fmt.Sprintf("attribute_not_exists(%s)", attributeLockId)),
+	}
+	_, err := client.PutItem(putParams)
+	if err != nil {
+		options.Logger.Errorf(
+			"Error acquiring lock %s in table %s in region %s (already locked?): %s\n",
+			options.LockString,
+			options.LockTable,
+			options.AwsRegion,
+			err,
+		)
+		return errors.WithStackTrace(err)
+	}
+	options.Logger.Infof("Acquired lock '%s' in table %s\n", options.LockString, options.LockTable)
+	return nil
+}
+
+// ReleaseLock will attempt to release the lock defined by the provided lock string in the configured lock table for the
+// configured region
+func ReleaseLock(options *Options) error {
+	client, err := NewDynamoDb(options.AwsRegion)
+	if err != nil {
+		options.Logger.Errorf("Error authenticating to AWS: %s\n", err)
+		return err
+	}
+
+	tableExists, err := lockTableExistsAndIsActive(options.LockTable, client)
+	if err != nil {
+		options.Logger.Errorf("Error checking if DynamoDB table %s exists and is active\n", options.LockTable)
+		return err
+	}
+
+	if !tableExists {
+		options.Logger.Errorf("DynamoDB table %s does not exist\n", options.LockTable)
+		return err
+	}
+
+	return releaseLockWithRetries(options, client)
+}
+
+// releaseLockWithRetries will attempt to release the lock defined by the provided lock string in the configured lock table
+// for the configured region. This will retry on failure, until reaching a timeout
+func releaseLockWithRetries(options *Options, client *dynamodb.DynamoDB) error {
+	return retry.DoWithRetry(
+		options.Logger,
+		fmt.Sprintf("Trying to release DynamoDB lock %s in table %s\n", options.LockString, options.LockTable),
+		options.MaxRetries,
+		options.SleepBetweenRetries,
+		func() error {
+			return releaseLock(options, client)
```

Comment:
I may be missing something, but why do we need to do the release lock in a retry loop? It makes sense to _acquire_ a lock in a retry loop, as you are waiting for someone else to release it, but shouldn't _release_ be a one-time action? Either you can release it, or something is wrong?

Reply:
Hm, you might be right actually! I thought it would make sense to retry it in case of an error, but we _should probably_ just throw the error.